derive-eip712/src/lib.rs

```diff
@@ -0,0 +1,399 @@
+    match field_type.as_ref() {
+        "U128" => "uint128",
+        "U256" => "uint256",
+        "H128" => "bytes16",
+        "H160" => "address",
+        "H256" => "bytes32",
+        "String" => "string",
+        "Bytes" => "bytes",
+        "Vec<u8>" => "bytes",
+        "Vec<U128>" => "uint128[]",
+        "Vec<U256>" => "uint256[]",
+        "Vec<H128?" => "bytes16[]",
+        "Vec<H160>" => "address[]",
+        "Vec<H256?" => "bytes32[]",
+        "Vec<String>" => "string[]",
+        "Vec<Bytes>" => "bytes[]",
+        _ => {
+            // NOTE: This will fail if the field type does not match an ethereum type;
+            &field_type
+        }
+    }
+    .to_string()
+}
+
+// Parse the field type from the derived struct
+fn parse_field(field: &syn::Field) -> String {
+    let field_path = match &field.ty {
+        syn::Type::Path(p) => p,
+        _ => {
+            panic!("field type must be a path")
+        }
+    };
+
+    let segment = field_path
+        .path
+        .segments
+        .first()
+        .expect("field must have a type");
+
+    let mut field_type = segment.ident.to_string();
+
+    if let syn::PathArguments::AngleBracketed(arguments) = &segment.arguments {
+        field_type.push('<');
+        for arg in &arguments.args {
+            if let syn::GenericArgument::Type(syn::Type::Path(p)) = arg {
+                let arg_identity = p
+                    .path
+                    .segments
+                    .first()
+                    .map(|s| s.ident.to_string())
+                    .expect("argument must have an identity");
+
+                field_type.push_str(&arg_identity);
+            }
+        }
+        field_type.push('>');
+    }
+
+    parse_field_type(field_type)
+}
+
+// Return HashMap of the field name and the field type;
+fn parse_fields(ast: &syn::DeriveInput) -> HashMap<String, String> {
+    let mut parsed_fields = HashMap::new();
+
+    let data = match &ast.data {
+        syn::Data::Struct(s) => s,
+        _ => {
+            panic!("Eip712 can only be derived for a struct")
+        }
+    };
+
+    let named_fields = match &data.fields {
+        syn::Fields::Named(name) => name,
+        _ => {
+            panic!("unnamed fields are not supported")
+        }
+    };
+
+    named_fields.named.iter().for_each(|f| {
+        let field_name = f
+            .ident
+            .clone()
+            .expect("field must be named")
+            .to_string()
+            .to_case(Case::Camel);
+
+        let field_type = parse_field(f);
+
+        parsed_fields.insert(field_name, field_type);
+    });
+
+    parsed_fields
+}
+
+// Convert hash map of field names and types into a type hash corresponding to enc types;
+fn make_type_hash(primary_type: String, fields: &HashMap<String, String>) -> String {
+    let parameters = fields
+        .iter()
+        .map(|(k, v)| format!("{} {}", v, k))
+        .collect::<Vec<String>>()
+        .join(",");
```

Comment:
Note to myself, these parameters should be sorted alphabetically.

Reply:
No longer using a hashmap, fields are sorted based on their position in the struct.